src/books/book.service.ts

```diff
@@ -0,0 +1,82 @@
+import { Injectable, NotFoundException } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Repository } from 'typeorm';
+import { BookEntity } from './book.entity';
+import { generateEAN13 } from 'src/utils/helpers';
+import { Book } from './domain/book.domain';
+import { BookCreate } from './domain/book-create';
+import { BookUpdate } from './domain/book-update';
+
+@Injectable()
+export class BookService {
+  constructor(
+    @InjectRepository(BookEntity)
+    private readonly bookRepository: Repository<BookEntity>,
+  ) {}
+
+  private async generateUniqueEAN13(): Promise<string> {
+    let ean: string;
+    let exists: BookEntity | null;
+
+    do {
+      ean = await generateEAN13();
+      exists = await this.bookRepository.findOne({ where: { code: ean } });
+    } while (exists);
+
+    return ean;
```

Comment:
Có thể viết gọn thế này. Chú ý improve

- Một biến lúc được khai báo thì luôn khai báo const
- Chọn hàm đúng nhất để sử dụng


```suggestion
    const code = await generateEAN13();
   const existing = await this.bookRepository.existingBy({ code });
   
   return existing ? this.generateUniqueEAN13(): code;
```